### PR TITLE
Blacklist

### PR DESCRIPTION
**Pull request by imbcmdth**
First shot at temporarily blacklisting problematic playlists. Will flag a playlist as bad and then select the next appropriate playlist. If no playlists remain, it will signal the error that caused the problem to begin with.

### CODE REVIEW

File: src/videojs-hls.js [hunk 1]
```diff
@@ -179,15 +180,7 @@ videojs.HlsHandler.prototype.src = function(src) {
...
-    // close the media source with the appropriate error type
-    if (this.playlists.error.code === 2) {
-      this.mediaSource.endOfStream('network');
-    } else if (this.playlists.error.code === 4) {
-      this.mediaSource.endOfStream('decode');
-    }
-
-    // if this error is unrecognized, pass it along to the tech
-    this.tech_.error(this.playlists.error);
+    this.blacklistCurrentPlaylist_(this.playlists.error.code);
```
Comment (dmlap) [inline src/videojs-hls.js]: Should we be passing the entire error object here?

File: src/videojs-hls.js [hunk 2]
```diff
@@ -916,6 +909,34 @@ videojs.HlsHandler.prototype.setBandwidth = function(xhr) {
   this.tech_.trigger('bandwidthupdate');
 };
 
+videojs.HlsHandler.prototype.blacklistCurrentPlaylist_ = function(error) {
+  var currentPlaylist, nextPlaylist;
+
+  currentPlaylist = this.playlists.media();
+
+  if (!currentPlaylist) {
```
Comment (dmlap) [inline src/videojs-hls.js]: When would this occur? (Maybe a comment would help here)

File: src/videojs-hls.js [hunk 3]
```diff
@@ -930,7 +951,11 @@ videojs.HlsHandler.prototype.loadSegment = function(segmentInfo) {
   this.segmentXhr_ = videojs.Hls.xhr({
     uri: segmentInfo.uri,
     responseType: 'arraybuffer',
-    withCredentials: this.source_.withCredentials
+    withCredentials: this.source_.withCredentials,
+    // Set xhr timeout to 150% of the segment duration to allow us
+    // some time to switch renditions in the event of a catastrophic
+    // decrease in network performance or a server issue.
+    timeout: (segment.duration * 1.5) * 1000
```
Comment (dmlap) [inline src/videojs-hls.js]: Nice. We shouldn't assume `segment.duration` is always defined, though-- how about a fallback to `targetDuration`?

File: test/videojs-hls_test.js [hunk 4]
```diff
@@ -1510,32 +1495,23 @@ test('playlist 404 should end stream with a network error', function() {
...
-
   openMediaSource(player);
 
-  standardXHRResponse(requests[0]);
-  requests[1].respond(404);
-  ok(player.tech_.hls.error.message, 'an error message is available');
-  equal(2, player.tech_.hls.error.code, 'Player error code should be set to MediaError.MEDIA_ERR_NETWORK');
-});
-
-test('segment 500 should trigger MEDIA_ERR_ABORTED', function () {
```
Comment (dmlap) [inline test/videojs-hls_test.js]: You kind of killed a test for 500 response codes here. Intentional?

**Comment by dmlap** [comment]
Looks good other than the couple questions/comments I put in there. It's awesome that adding support for redundant streams turned out to be a relatively tiny change!

File: test/videojs-hls_test.js (hunk 4 above)
Comment (imbcmdth) [inline test/videojs-hls_test.js]: All non-200 responses are handled by exactly the same code. I didn't see the point of testing each individual code! :D

**Comment by imbcmdth** [comment]
> It's awesome that adding support for redundant streams turned out to be a relatively tiny change! 

Even WITH blacklisting timeouts!  :wink:

**Pull request merged**